### DISCUSSION TITLE
Passing an existing Logger object to the Roda application

### DISCUSSION THREAD

**Discussion by jiggneshhgohel** [jeremyevans/roda#238] [General]
**Background**

The background for this discussion can be found between comments https://github.com/jeremyevans/roda/issues/237#issuecomment-1011176605 and https://github.com/jeremyevans/roda/issues/237#issuecomment-1011270516

**Question**

Instead of a Roda application instantiating a Logger inside the application, is it possible to pass an already instantiated Logger to the Roda application? For an example use-case please refer the code in https://github.com/jeremyevans/roda/issues/237#issuecomment-1011215985. 

@jeremyevans Trying out that code your suspection in your comment https://github.com/jeremyevans/roda/issues/237#issuecomment-1011269617 is correct. I created a file `my_plugin.rb` at the level of `app.rb` itself and in it added following content

```
# frozen-string-literal: true

module MyPlugin
  #def self.configure(app, logger=nil, opts=OPTS)
  def self.configure(app, logger=nil, opts={})
    puts ">>>> #{logger.inspect}"

    app.opts[:common_logger] = logger || app.opts[:common_logger] || $stderr
    app.opts[:common_logger_meth] = app.opts[:common_logger].method(opts.fetch(:method){logger.respond_to?(:write) ? :write : :<<})
  end

  if RUBY_VERSION >= '2.1'
    # A timer object for calculating elapsed time.
    def self.start_timer
      Process.clock_gettime(Process::CLOCK_MONOTONIC)
    end
  else
    # :nocov:
    def self.start_timer # :nodoc:
      Time.now
    end
    # :nocov:
  end

  module InstanceMethods
    private

    # Log request/response information in common log format to logger.
    def _roda_after_90__common_logger(result)
      return unless result && result[0] && result[1]

      elapsed_time = if timer = @_request_timer
        '%0.4f' % (MyPlugin.start_timer - timer)
      else
        '-'
      end

      env = @_request.env

      opts[:common_logger_meth].call("#{env['HTTP_X_FORWARDED_FOR'] || env["REMOTE_ADDR"] || "-"} - #{env["REMOTE_USER"] || "-"} [#{Time.now.strftime("%d/%b/%Y:%H:%M:%S %z")}] \"#{env["REQUEST_METHOD"]} #{env["SCRIPT_NAME"]}#{env["PATH_INFO"]}#{"?#{env["QUERY_STRING"]}" if ((qs = env["QUERY_STRING"]) && !qs.empty?)} #{env["HTTP_VERSION"]}\" #{result[0]} #{((length = result[1]['Content-Length']) && (length unless length == '0')) || '-'} #{elapsed_time}\n")
    end

    # Create timer instance used for timing
    def _roda_before_05__common_logger
      @_request_timer = MyPlugin.start_timer
    end
  end
end
```

Then in my `app.rb`

```
require "roda"
require_relative './my_plugin'

class Api < Roda

  class << self
     def set_app_logger=(logger)
       @app_logger = logger
     end

     def app_logger
        @app_logger
     end
  end

  plugin MyPlugin, self.app_logger

 ...
 ...
end

class App < Roda
  route do |r|
    # Reference: http://roda.jeremyevans.net/rdoc/files/README_rdoc.html#label-Composition
    r.on 'api' do
      r.run Api
    end
  end
end
```

Then when I started the application using `rackup` in terminal following was the output

```
>>>> nil
[2022-01-13 15:37:45] INFO  WEBrick 1.7.0
[2022-01-13 15:37:45] INFO  ruby 3.0.2 (2021-07-07) [x86_64-linux]
[2022-01-13 15:37:45] INFO  WEBrick::HTTPServer#start: pid=5596 port=9292
```

and the `Rack::Unreloader` logs were written to `logs/development.log` file in my application but the request logs were still written to the terminal which is obvious because the plugin was not passed the logger created and set in `config.ru`.

So is there any provision to make this work in Roda-way instead of some workarounds or patch? I think with a concept like `load_plugins_after_app_loads` or `lazy_load_plugins` can open up a way to make this work but those are just vague thoughts and how much practical and useful those ideas can be, that only you can comment on.

Please share your thoughts.

Thanks.

**Comment by jeremyevans** [comment]
It seems the easiest way to handle this would be to set the logger to a constant in config.ru (`logger = LOGGER =  Logger.new(env_log_file_path)`), and then use that in your app.rb (`plugin MyPlugin, LOGGER`).

**Comment by jiggneshhgohel** [comment]
@jeremyevans That worked. Thanks a lot for your prompt help.
